nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationLifecycle.java

```diff
@@ -196,6 +197,8 @@ public class ApplicationLifecycle implements Deployment, PostConstruct {
     private ExecutorService executorService;
     private Collection<ApplicationLifecycleInterceptor> alcInterceptors = emptyList();
 
+    private ThreadLocal<Deque<ExtendedDeploymentContext>> currentDeploymentContext;
```

Comment:
How are you ensuring cleanup on failed deployments? I think we had some similar leak at some point in the past ...

Reply:
After each `pushCurrentDeploymentContext()` call, there's `popCurrentDeploymentContext()` call in a finally block inside the same method. The last call to `popCurrentDeploymentContext` will remove the thread local variable.